src/robot_state_publisher.cpp

```diff
@@ -68,110 +73,308 @@ geometry_msgs::msg::TransformStamped kdlToTransform(const KDL::Frame & k)
   return t;
 }
 
-RobotStatePublisher::RobotStatePublisher(
-  rclcpp::Node::SharedPtr node_handle,
-  const KDL::Tree & tree,
-  const urdf::Model & model,
-  const std::string & model_xml)
-: model_(model),
-  tf_broadcaster_(node_handle),
-  static_tf_broadcaster_(node_handle),
-  clock_(node_handle->get_clock())
+}  // namespace
+
+RobotStatePublisher::RobotStatePublisher(const rclcpp::NodeOptions & options)
+: rclcpp::Node("robot_state_publisher", options)
 {
-  // walk the tree and add segments to segments_
-  addChildren(tree.getRootSegment());
+  // get the XML
+  std::string urdf_xml = this->declare_parameter("robot_description", std::string(""));
+  if (urdf_xml.empty()) {
+    throw std::runtime_error("robot_description parameter must be provided");
+  }
+
+  // set publish frequency
+  double publish_freq = this->declare_parameter("publish_frequency", 20.0);
+  if (publish_freq > 1000.0) {
+    throw std::runtime_error("publish_frequency must be <= 1000.0");
```

Comment:
Why is this a constraint? Seems arbitrary.

Reply:
It's because of the math below.  If the user specified 2000.0 for instance, they would end up dividing 1000.0 / 2000.0, which gets them 0.5 in double, but which ends up as 0 when cast to a uint64_t .  We could allow publishing at faster than 1000Hz by changing the type from `std::chrono::milliseconds` -> `std::chrono::microseconds`, but I didn't really see the point.

Reply:
Okay. It does seem unlikely someone will run into this constraint, and if they do we can always change to microseconds or nanoseconds.

We should probably guard against negative numbers too.

Reply:
Good point.  Fixed in db174336a159e586eb0dddbb2d6c3303fc610df4.